libclc/CMakeLists.txt

```diff
@@ -1,47 +1,110 @@
 
-# If we are not building as a part of LLVM, build libclc as an
-# standalone project, using LLVM/Clang as external tools.
-if( CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR )
-  project( libclc VERSION 0.2.0 LANGUAGES CXX C )
+set(CMAKE_CXX_STANDARD 17)
 
-  set( LIBCLC_STANDALONE_BUILD 1 )
+# Add path for custom modules
+list( INSERT CMAKE_MODULE_PATH 0 "${PROJECT_SOURCE_DIR}/cmake/modules" )
 
-  include(HandleOutOfTreeLLVM)
-else()
-  include(HandleInLLVMTree)
-endif()
+set( LIBCLC_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR} )
+set( LIBCLC_BINARY_DIR ${CMAKE_CURRENT_BINARY_DIR} )
+set( LIBCLC_OBJFILE_DIR ${LIBCLC_BINARY_DIR}/obj.libclc.dir )
 
-set(CMAKE_CXX_STANDARD 17)
+include( AddLibclc )
 
 include( GNUInstallDirs )
 set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS
   amdgcn-amdhsa/lib/SOURCES;
+  amdgcn-amdhsa/libspirv/SOURCES;
   amdgcn/lib/SOURCES;
+  amdgcn/libspirv/SOURCES;
   amdgcn-mesa3d/lib/SOURCES;
   amdgpu/lib/SOURCES;
   clspv/lib/SOURCES;
   clspv64/lib/SOURCES;
   generic/lib/SOURCES;
+  generic/libspirv/SOURCES;
   ptx/lib/SOURCES;
   ptx-nvidiacl/lib/SOURCES;
+  ptx-nvidiacl/libspirv/SOURCES;
   r600/lib/SOURCES;
+  r600/libspirv/SOURCES;
   spirv/lib/SOURCES;
   spirv64/lib/SOURCES
+  x86_64-unknown-linux/libspirv/SOURCES
 )
 
-include( AddLibclc )
+set( LIBCLC_MIN_LLVM 3.9.0 )
+
+set( LIBCLC_TARGETS_TO_BUILD "all"
+    CACHE STRING "Semicolon-separated list of targets to build, or 'all'." )
+
+option( ENABLE_RUNTIME_SUBNORMAL "Enable runtime linking of subnormal support." OFF )
+
+if( LIBCLC_STANDALONE_BUILD OR CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR )
+  # Out-of-tree configuration
+  set( LIBCLC_STANDALONE_BUILD TRUE )
+
+  find_package(LLVM REQUIRED HINTS "${LLVM_CMAKE_DIR}")
+  include(AddLLVM)
+
+  message( STATUS "libclc LLVM version: ${LLVM_PACKAGE_VERSION}" )
+
+  if( LLVM_PACKAGE_VERSION VERSION_LESS LIBCLC_MIN_LLVM )
+    message( FATAL_ERROR "libclc needs at least LLVM ${LIBCLC_MIN_LLVM}" )
+  endif()
+
+  # Import required tools as targets
+  if ( NOT EXISTS LIBCLC_CUSTOM_LLVM_TOOLS_BINARY_DIR )
+    foreach( tool clang llvm-as llvm-link opt )
+      find_program( LLVM_TOOL_${tool} ${tool} PATHS ${LLVM_TOOLS_BINARY_DIR} NO_DEFAULT_PATH )
+      add_executable( libclc::${tool} IMPORTED GLOBAL )
+      set_target_properties( libclc::${tool} PROPERTIES IMPORTED_LOCATION ${LLVM_TOOL_${tool}} )
+    endforeach()
+  endif()
+else()
+  # In-tree configuration
+  set( LIBCLC_STANDALONE_BUILD FALSE )
 
-# List of all targets
+  set( LLVM_PACKAGE_VERSION ${LLVM_VERSION} )
+
+  # Note that we check this later (for both build types) but we can provide a
+  # more useful error message when built in-tree. We assume that LLVM tools are
+  # always available so don't warn here.
+  if( NOT clang IN_LIST LLVM_ENABLE_PROJECTS )
+    message(FATAL_ERROR "Clang is not enabled, but is required to build libclc in-tree")
+  endif()
+
+  if ( NOT EXISTS LIBCLC_CUSTOM_LLVM_TOOLS_BINARY_DIR )
```

Comment:
There's basically an `else` clause below, but it's a standalone if clause gaurded by the opposite condition. Perhaps you could be a diamond geezer, and turn this into an if-else?

Reply:
fair